src/main/java/com/google/sps/data/PersistentUserRepository.java

```diff
@@ -0,0 +1,130 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.User;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+/**
+ * This class implements the match repository using the datastore 
+ */
+public class PersistentUserRepository implements UserRepository {
+    
+  private Collection<User> allUsers = new ArrayList<>();
+  
+  private DatastoreService datastore;
+  
+  private int maxProfiles = 10;
+  
+  public PersistentUserRepository() {
+      datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  // function to add profile to database
+  public void addUserToDatabase(User user) {
+    String name = user.toString();
+    Collection<String> specialties = user.getSpecialties();
+
+    Entity userEntity = new Entity("User");
+    userEntity.setProperty("name", name);
+    if(specialties.size() > 0){
+        userEntity.setProperty("specialties", specialties);
+    }
+    datastore.put(userEntity);
+  }
+
+  // function to fetch the user from the database
+  public Collection<User> fetchUserProfiles() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    Collection<User> userProfiles = new ArrayList<>();
+    for (Entity entity : resultsList) {
+        String name = (String) entity.getProperty("name");
+        Collection<String> specialties = (Collection) entity.getProperty("specialties");
+
+        User userObject = new User(name);
+        if(specialties != null) {
+            for (String specialty : specialties) {
+                userObject.addSpecialty(specialty);
+            }
+        }
+        userProfiles.add(userObject);
+    }
+    return userProfiles;
+  }
+
+  public void removeUserProfile(User user){
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    String name = user.toString();
+
+    for (Entity entity : resultsList) {
+        String currentName = (String) entity.getProperty("name");
+        if(name.contains(currentName)) {
```

Comment:
If I have a user who works at mckinsey and another user who's name is mckinsey, would this code delete both users?

Reply:
It would not conflict because remove user right now is just checking the name attribute, not the company property. So it would not conflict

Reply:
Recall that user.toString() returns the json string which contains more than just the name. It will contain the company, major, occupation, and school as well.

Reply:
Right, but when we remove users, I'd look more specifically at their name property because that's what we are looking at specifically.

Reply:
when i all .getName() method that returns a User's name, not the json string which is what I use to compare for deleting

Reply:
Done.